packages/apollo-federation/src/composition/compose.ts

```diff
@@ -485,31 +510,38 @@ export function addFederationMetadataToSchemaNodes({
     const namedType = schema.getType(field.parentTypeName);
     if (!namedType) continue;
 
-    namedType.federation = {
-      ...namedType.federation,
+    const existingMetadata = getFederationMetadata(namedType);
+    const typeFederationMetadata: FederationType = {
+      ...existingMetadata,
       externals: {
-        ...(namedType.federation && namedType.federation.externals),
+        ...existingMetadata?.externals,
         [field.serviceName]: [
-          ...(namedType.federation &&
-          namedType.federation.externals &&
-          namedType.federation.externals[field.serviceName]
-            ? namedType.federation.externals[field.serviceName]
-            : []),
+          ...(existingMetadata?.externals?.[field.serviceName] || []),
           field,
         ],
       },
     };
+
+    namedType.extensions = {
+      ...namedType.extensions,
+      federation: typeFederationMetadata,
+    };
   }
 
   // add all definitions of a specific directive for validation later
   for (const directiveName of Object.keys(directiveDefinitionsMap)) {
     const directive = schema.getDirective(directiveName);
     if (!directive) continue;
 
-    directive.federation = {
-      ...directive.federation,
+    const directiveFederationMetadata: FederationDirective = {
+      ...getFederationMetadata(directive),
       directiveDefinitions: directiveDefinitionsMap[directiveName],
-    };
+    }
+
+    directive.extensions = {
+      ...directive.extensions,
+      federation: directiveFederationMetadata,
+    }
```

Comment:
👻

Reply:
@IvanGoncharov jut pointing out one more place we didn't address in our discussion: directive extensions!